File: auto_outsrc/parser/ARCHIVE/decOutOutsourcing_BSW.cpp
#include "sdlconfig.h"
#include <iostream>
#include <sstream>
#include <string>
using namespace std;
#define DEBUG  true

string decout(PairingGroup & groupObj, CharmDict & partCT, ZR & zz, GT & egg)
{
	CharmDict input;
	input.append(partCT);
	input.append(zz);
	input.append(egg);
	CharmDict partCT;
	partCT.append(T0);
	partCT.append(T1);
	partCT.append(T2);
	GT R = groupObj.div(T0, groupObj.exp(T2, zz));
	string s_sesskey = DeriveKey(R);
	string M = SymDec(s_sesskey, T1);
	CharmDict hashRandM;
	hashRandM.append(R);
	hashRandM.append(M);
	ZR s = groupObj.hashListToZR(hashRandM);
	if ( ( (( (T0) == (groupObj.mul(R, groupObj.exp(egg, s))) )) && (( (T2) == (groupObj.exp(egg, groupObj.div(s, zz))) )) ) )
	{
		CharmDict output = M;
	}
	else
	{
		userErrorFunction('invalid ciphertext');
		return NULL;
	}
	return M;
}

if __name__ == "__main__":
	global groupObj
	groupObj = PairingGroup('SS512')

	M = decout(sys.argv[1], sys.argv[2], sys.argv[3])

File: auto_outsrc/parser/ARCHIVE/BSW07/decOutOutsourcing_BSW.py
from userFuncs_BSW import *
from toolbox.pairinggroup import PairingGroup, ZR, G1, G2, GT, pair
from toolbox.secretutil import SecretUtil
from toolbox.symcrypto import AuthenticatedCryptoAbstraction
from toolbox.iterate import dotprod2
from charm.pairing import hash as SHA1

def decout(partCT, zz, egg):
	input = [partCT, zz, egg]
	T0, T1, T2 = partCT
	R = (T0 / (T2 ** zz))
	s_sesskey = SHA1(R)
	M = SymDec(s_sesskey, T1)
	s = groupObj.hash([R, M], ZR)
	output = M
	return output

if __name__ == "__main__":
	global groupObj
	groupObj = PairingGroup('SS512')

	(M) = decout(sys.argv[1], sys.argv[2], sys.argv[3])
